pkg/apis/softwarecomposition/network_types.go

```diff
@@ -0,0 +1,61 @@
+package softwarecomposition
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Protocol string
+type CommunicationType string
+
+const (
+	ProtocolTCP  Protocol = "TCP"
+	ProtocolUDP  Protocol = "UDP"
+	ProtocolSCTP Protocol = "SCTP"
+
+	CommunicationTypeIngress CommunicationType = "internal"
+	CommunicationTypeEgress  CommunicationType = "external"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NetworkNeighborsList is a list of NetworkNeighbors.
+type NetworkNeighborsList struct {
+	metav1.TypeMeta
+	metav1.ListMeta
+
+	Items []NetworkNeighbors
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NetworkNeighbors represents a list of network communications for a specific workload.
+type NetworkNeighbors struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec NetworkNeighborsSpec
+}
+
+type NetworkNeighborsSpec struct {
+	*metav1.LabelSelector // The labels which are inside spec.selector in the parent workload.
+	Ingress               []NetworkEntry
+	Egress                []NetworkEntry
+}
+
+// IngressEntry represents a single network communication.
+type NetworkEntry struct {
+	Identifier        string
+	Type              CommunicationType
+	DNS               string
+	Ports             []NetworkPort
+	PodSelector       *metav1.LabelSelector
+	NamespaceSelector *metav1.LabelSelector
+	IPAddress         string
+}
+
+type NetworkPort struct {
+	Name     string // protocol-port
+	Protocol Protocol
+	Port     uint16
```

Comment:
When should `uint` be used if not for positive values?

Reply:
I’m talking more about size of the type, not its sign. But now that I look at how [Deployments](https://github.com/kubernetes/api/blob/fe172d7dd4b6501923938705cfb69ee0b907c32d/apps/v1/types.go#L376) and [Flow Control](https://github.com/kubernetes/api/blob/fe172d7dd4b6501923938705cfb69ee0b907c32d/flowcontrol/v1beta3/types.go#L155) are implemented. It seems that it’s a convention to specify the size, but avoid `uint` even for non-negative values.

I believe this is because of serialization, so that overflowing values don’t get silently cast to `uint`, and instead get handled at the validator. For example, if you pass `replicas: -1`, a server will be able to deserialize it properly, then validate the value and give you a better, meaningful error.

Reply:
I see. I actually made it based on the type that `inspektor gadget` uses at the event they create, but you are right